### DISCUSSION TITLE
Duplicate Middleware Execution for Overlapping Command and Text Handlers

### DISCUSSION THREAD

**Discussion by sr-44** [nutgram/nutgram#758] [Q&A] [answered]
If there's a handler like $bot->onCommand('/start'); and another handler $bot->onText('.*');, a user's message is first processed by the first handler and then by the second handler. Although I can exclude the processing of commands in the second handler through conditions, but the middlewares are executed twice because of this. For example:

```php
$bot->onCommand('start', StartCommand::class);
$bot->onText('.*', OtherCommand::class);
``` 

```php
public function __invoke(Nutgram $bot, $next)
{
    if (someCondition()){
        $bot->sendMessage('some message');
        return;
    }
    $next($bot);
}
``` 

When I start, I receive 'some message' twice.

Is this behavior intended, or is this something that shouldn't happen? What should I do to avoid this?

**Comment by Lukasss93** [comment]
We might think about implementing a feature that fixes this “problem” ...
I'll try to make a list of ideas that come to mind:

1. `$bot->onListenerXXX(...)->ignoreIfPatternIsCommand();` + `$bot->group(...)->ignoreIfPatternIsCommand();` (it doesn't work with `$bot->onCommand(...)` or `$bot->registerCommand(...)`) 🤮
2. Add a flag to Configuration class ❓
3. Hardcoded fix (BC) ❗

@sergix44 what do you think? do you have any ideas/suggestions on how we can handle this problem?

**Reply by sergix44** [reply]
Why?
This is intended behaviour, since the update is going through the handler stack twice.

https://nutgram.dev/docs/usage/middleware#skipping-global-middlewares-for-a-specific-handler

This should be more than enough applied to just the start command

**Reply by sergix44** [reply]
Or, if you are calling the handler middlewares, is another oneliner
`$bot->onText('^(?!\/start\b).+', OtherCommand::class);
`

**Reply by Lukasss93** [reply]
You're right. 😅

Or you can create a middleware and using this (undocumented) proxy: $bot->isCommand()

```php
class NoCommands
{
    public function __invoke(Nutgram $bot, $next, CustomService $service): void
    {
        if($bot->isCommand()) {
            return;
        }

        $next($bot);
    }
}
```

```php
$bot->onText('.*', OtherCommand::class)->middleware(NoCommands::class);
```

**Reply by sr-44** [reply]
I had already thought of that, but it didn't work.

**Comment by sr-44** [comment, marked as answer]
Instead of `onText` I tried `onMessageType(MessageType::TEXT)` and so far no problems, everything works
